pkg/infra/remotecache/ring/cache.go

```diff
@@ -0,0 +1,233 @@
+package ring
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"hash/fnv"
+	"net"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/grafana/dskit/kv"
+	"github.com/grafana/dskit/kv/memberlist"
+	"github.com/grafana/dskit/ring"
+	"github.com/grafana/dskit/services"
+	"github.com/prometheus/client_golang/prometheus"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/infra/remotecache/common"
+	"github.com/grafana/grafana/pkg/services/grpcserver"
+	"github.com/grafana/grafana/pkg/setting"
+)
+
+const CacheType = "ring"
+
+func NewCache(cfg *setting.Cfg, reg prometheus.Registerer, provider grpcserver.Provider) (*Cache, error) {
+	logger := log.New("remotecache.ring")
+
+	grpcAddr, grpcPort, err := net.SplitHostPort(cfg.GRPCServerAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	advertiseAddr := cfg.RemoteCache.Ring.Addr
+	// fallback to grpc address
+	if advertiseAddr == "" {
+		advertiseAddr = grpcAddr
+	}
+
+	memberlistsvc, client, err := newMemberlistService(memberlistConfig{
+		AdvertiseAddr: advertiseAddr,
+		AdvertisePort: cfg.RemoteCache.Ring.Port,
+		Addr:          grpcAddr,
+		Port:          cfg.RemoteCache.Ring.Port,
```

Comment:
I don't understand the way we're mixing grpcAddr, advertiseAddr, grpcPort & Ring.Port here and in newRing below. It seems like in different places we use different combinations.

Reply:
I has been a while since I worked on this but IIRC it is something like this:

The member list service will start its own grpc server and we cannot control this so we need to use a different port for that than the port we use for the internal grpc server. So port + addr is what is started up locally. The advertise address and port is how other nodes can reach this node. If we are running two nodes on different machines they need to know these so they can communicate with it.

The configuration passed in the ring is just metadata that other nodes can read. Because we are adding the grpc service `Dispatcher` to our general grpc server use the advertiseAddr + grpc port. This is how different nodes can dispatch cache requests to other nodes.

I believe I added this snippet for convenience when trying it out locally:
```go
advertiseAddr := cfg.RemoteCache.Ring.Addr
// fallback to grpc address
if advertiseAddr == "" {
	advertiseAddr = grpcAddr
}
``` 

If I start node 1 on `127.0.0.1` and node 2 on `127.0.0.2` these can also be used as their advertise address because we are running both nodes on the same machine.

Does this make sense to you?